### PR TITLE
online word2vec

### PR DESCRIPTION
**Pull request by isohyt**
Rebase of #778 .
It's ready to merge :)

### CODE REVIEW

File: gensim/models/word2vec.py [hunk 1]
```diff
@@ -577,26 +577,44 @@ def scale_vocab(self, min_count=None, sample=None, dry_run=False, keep_raw_vocab
...
+            for word, v in iteritems(self.raw_vocab):
+                if keep_vocab_item(word, v, min_count, trim_rule=trim_rule):
+                    retain_words.append(word)
+                    retain_total += v
+                    if not dry_run:
+                        if word in self.vocab:
+                            self.vocab[word].count += v
+                        else:
+                            self.vocab[word] = Vocab(count=v, index=len(self.index2word))
+                            self.index2word.append(word)
```
Comment (tmylk) [inline gensim/models/word2vec.py]: Could you please add logging for the number of new words added to vocab?
```
logger.info("New words added to vocab: %d ", new_word_count)
```

File: gensim/test/test_word2vec.py [hunk 2]
```diff
@@ -110,30 +175,6 @@ def testPersistenceWord2VecFormat(self):
         norm_only_model.init_sims(replace=True)
         self.assertFalse(numpy.allclose(model['human'], norm_only_model['human']))
         self.assertTrue(numpy.allclose(model.syn0norm[model.vocab['human'].index], norm_only_model['human']))
```
Comment (tmylk) [inline gensim/test/test_word2vec.py]: why delete the lines below?

File: gensim/test/test_word2vec.py [hunk 3]
```diff
@@ -370,12 +411,9 @@ def testSimilarities(self):
         model = word2vec.Word2Vec(size=2, min_count=1, sg=0, hs=0, negative=2)
         model.build_vocab(sentences)
         model.train(sentences)
-        
+
         self.assertTrue(model.n_similarity(['graph', 'trees'], ['trees', 'graph']))
         self.assertTrue(model.n_similarity(['graph'], ['trees']) == model.similarity('graph', 'trees'))
-        self.assertRaises(ZeroDivisionError, model.n_similarity, ['graph', 'trees'], [])
```
Comment (tmylk) [inline gensim/test/test_word2vec.py]: please only have your changes

**Comment by tmylk** [comment]
Please add logging and redo the merge. 

We are almost done with this needed feature!

**Comment by isohyt** [comment]
Oops, I overlooked test_w2v compatibility.
I will fix these which you pointed out within the week.

File: gensim/corpora/wikicorpus.py [hunk 4]
```diff
@@ -214,9 +214,10 @@ def extract_pages(f, filter_namespaces=False):
             title = elem.find(title_path).text
             text = elem.find(text_path).text
 
-            ns = elem.find(ns_path).text
```
Comment (tmylk) [inline gensim/corpora/wikicorpus.py]: @isohyt  why is this file here?

File: gensim/models/word2vec.py (hunk 1 above)
Comment (tmylk) [inline gensim/models/word2vec.py]: Ping @isohyt

File: gensim/corpora/wikicorpus.py (hunk 4 above)
Comment (isohyt) [inline gensim/corpora/wikicorpus.py]: When I transform old enwiki dump (2010 in this notebook), Attribution Error was returned as follows.

`  File "***/lib/python3.5/site-packages/gensim/corpora/wikicorpus.py", line 211, in extract_pages
    ns = elem.find(ns_path).text
AttributeError: 'NoneType' object has no attribute 'text'`

To avoid this error, I fixed wikicorpus.py.

**Comment by tmylk** [comment]
What about logging of the count of new words?

File: gensim/models/word2vec.py (hunk 1 above)
Comment (isohyt) [inline gensim/models/word2vec.py]: sorry i forgot it.
it's added in code

**Comment by isohyt** [comment]
hmm, test_w2v.py was not passed that check the similarity the word between 'war' and 'terrorist'  just python version 2.X.
is this test too severe for small split lee corpus?

File: gensim/models/word2vec.py [hunk 5]
```diff
@@ -577,34 +577,66 @@ def scale_vocab(self, min_count=None, sample=None, dry_run=False, keep_raw_vocab
...
+                        if not dry_run:
+                            self.vocab[word] = Vocab(count=v, index=len(self.index2word))
+                            self.index2word.append(word)
+                else:
+                    drop_unique += 1
+                    drop_total += v
+            original_unique_total = len(pre_exist_words) + len(new_words) + drop_unique
+            pre_exist_unique_pct = len(pre_exist_words) * 100 / max(original_unique_total, 1)
+            new_unique_pct = len(new_words) * 100 / max(original_unique_total, 1)
+            logger.info("""New added %i unique words (%i%% of original %i)
```
Comment (tmylk) [inline gensim/models/word2vec.py]: Sorry, I don't understand this logger message. could you please add an example in the notebook?

**Comment by tmylk** [comment]
Please add the logging example to the notebook on a small sample. It is hard to understand what is happening with the current messages.

File: gensim/models/word2vec.py [hunk 6]
```diff
@@ -577,34 +577,66 @@ def scale_vocab(self, min_count=None, sample=None, dry_run=False, keep_raw_vocab
...
+                        and increased the count of %i pre-existing words (%i%% of original %i)""",
+                        len(new_words), new_unique_pct, original_unique_total,
+                        len(pre_exist_words), pre_exist_unique_pct, original_unique_total)
+            retain_words = new_words + pre_exist_words
+            retain_total = new_total + pre_exist_total
```
Comment (isohyt) [inline gensim/models/word2vec.py]: retain_words and retain_total are used for the down sampling.
They are mainly used from line 642 to 663

File: gensim/models/word2vec.py (hunk 5 above)
Comment (isohyt) [inline gensim/models/word2vec.py]: you mean to add this logger on online_w2v_tutorial.ipynb?

Comment (tmylk) [inline gensim/models/word2vec.py]: Better add to the main notebook.

**Comment by tmylk** [comment]
@isohyt Thanks for finishing this code!
Let's update the notebook in a separate PR.

**Comment by isohyt** [comment]
thanks for staying with me on this long journey, @tmylk :)
i will do that asap

**Comment by piskvorky** [comment]
@isohyt Has the notebook been updated?

I still see
```
This implementation is still beta version at 16/09/04. You can download the beta version of online word2vec implementation in the following repository.
In [ ]:
%%bash
git clone -b online-w2v git@github.com:isohyt/gensim.git
```
at the top of the notebook. Can you please update?

@tmylk what else needs to be done for this to be complete?

**Comment by tmylk** [comment]
@isohyt the line about 'beta' and link to your private github repo should be removed. This project is complete.

**Comment by isohyt** [comment]
Sorry, I forgot to remove that description.
I will update this notebook asap in other PR.

**Comment by isohyt** [comment]
this notebook has been already fixed in dev branch now.
Thanks for fixing @tmylk !
